File: src/main/java/net/jgp/books/spark/ch17/lab200_feed_delta/FeedDeltaLakeApp.java
package net.jgp.books.spark.ch17.lab200_feed_delta;

import static org.apache.spark.sql.functions.col;
import static org.apache.spark.sql.functions.expr;
import static org.apache.spark.sql.functions.when;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

/**
 * Ingestion the 'Grand Débat' files to Delta Lake.
 * 
 * @author jgp
 */
public class FeedDeltaLakeApp {

  /**
   * main() is your entry point to the application.
   * 
   * @param args
   */
  public static void main(String[] args) {
    FeedDeltaLakeApp app =
        new FeedDeltaLakeApp();
    app.start();
  }

  /**
   * The processing code.
   */
  private void start() {
    // Create a session on a local master
    SparkSession spark = SparkSession.builder()
        .appName("Ingestion the 'Grand Débat' files to Delta Lake")
        .master("local[*]")
        .getOrCreate();

    // Create the schema
    StructType schema = DataTypes.createStructType(new StructField[] {
        DataTypes.createStructField(
            "authorId",
            DataTypes.StringType,
            false),
        DataTypes.createStructField(
            "authorType",
            DataTypes.StringType,
            true),
        DataTypes.createStructField(
            "authorZipCode",
            DataTypes.StringType,
            true),
        DataTypes.createStructField(
            "body",
            DataTypes.StringType,
            true),
        DataTypes.createStructField(
            "createdAt",
            DataTypes.TimestampType,
            false),
        DataTypes.createStructField(
            "enabled",
            DataTypes.BooleanType,
            true),
        DataTypes.createStructField(
            "endAt",
            DataTypes.TimestampType,
            true),
        DataTypes.createStructField(
            "fullAddress",
            DataTypes.StringType,
            true),
        DataTypes.createStructField(
            "id",
            DataTypes.StringType,
            false),
        DataTypes.createStructField(
            "lat",
            DataTypes.DoubleType,
            true),
        DataTypes.createStructField(
            "link",
            DataTypes.StringType,
            true),
        DataTypes.createStructField(
            "lng",
            DataTypes.DoubleType,
            true),
        DataTypes.createStructField(
            "startAt",
            DataTypes.TimestampType,
            false),
        DataTypes.createStructField(
            "title",
            DataTypes.StringType,
            true),
        DataTypes.createStructField(
            "updatedAt",
            DataTypes.TimestampType,
            true),
        DataTypes.createStructField(
            "url",
            DataTypes.StringType,
            true) });

    // Reads a JSON file, called 20190302 EVENTS.json, stores it in a
    // dataframe
    Dataset<Row> df = spark.read().format("json")
        .schema(schema)
        .option("timestampFormat", "yyyy-MM-dd HH:mm:ss")
        .load("data/france_grand_debat/20190302 EVENTS.json");

    df = df
        .withColumn("authorZipCode",
            col("authorZipCode").cast(DataTypes.IntegerType))
        .withColumn("authorZipCode",
            when(col("authorZipCode").$less(1000), null)
                .otherwise(col("authorZipCode")))
        .withColumn("authorZipCode",
            when(col("authorZipCode").$greater$eq(99999), null)
                .otherwise(col("authorZipCode")))
        .withColumn("authorDept", expr("int(authorZipCode / 1000)"));
    df.show(25);
    df.printSchema();

    df.write().format("delta")
        .mode("overwrite")
        .save("/tmp/delta_grand_debat_events");

    System.out.println(df.count() + " rows updated.");
  }
}

File: src/main/python/lab200_feed_delta/feedDeltaLakeApp.py
"""
 Ingestion the 'Grand Debate' files to Delta Lake.

 @author rambabu.posa
"""
import logging
import os
from pyspark.sql import (SparkSession, functions as F)
from pyspark.sql.types import (StructType, StructField, StringType,
                               TimestampType, DoubleType, BooleanType)

# Use this if you face any 'UnicodeEncodeError' error
#import sys
#reload(sys)
#sys.setdefaultencoding('utf-8')

def get_absolute_file_path(path, filename):
    current_dir = os.path.dirname(__file__)
    relative_path = f"{path}{filename}"
    absolute_file_path = os.path.join(current_dir, relative_path)
    return absolute_file_path

def main(spark):
    path = "../../../../data/france_grand_debat/"
    filename = "20190302 EVENTS.json"
    absolute_file_path = get_absolute_file_path(path, filename)

    # Create the schema
    schema = StructType([StructField('authorId', StringType(), False),
                         StructField('authorType', StringType(), True),
                         StructField('authorZipCode', StringType(), True),
                         StructField('body', StringType(), True),
                         StructField('createdAt', TimestampType(), False),
                         StructField('enabled', BooleanType(), True),
                         StructField('endAt', TimestampType(), True),
                         StructField('fullAddress', StringType(), True),
                         StructField('id', StringType(), False),
                         StructField('lat', DoubleType(), True),
                         StructField('link', StringType(), True),
                         StructField('lng', DoubleType(), True),
                         StructField('startAt', TimestampType(), False),
                         StructField('title', StringType(), True),
                         StructField('updatedAt', TimestampType(), True),
                         StructField('url', StringType(), True)])

    # Reads a JSON file, called 20190302 EVENTS.json,
    # stores it in a dataframe
    df = spark.read.format("json") \
        .schema(schema) \
        .option("timestampFormat", "yyyy-MM-dd HH:mm:ss") \
        .load(absolute_file_path)

    df = df.withColumn("authorZipCode", F.col("authorZipCode").cast("int")) \
        .withColumn("authorZipCode", F.when(F.col("authorZipCode")< F.lit(1000), None).otherwise(F.col("authorZipCode"))) \
        .withColumn("authorZipCode", F.when(F.col("authorZipCode") >= F.lit(99999), None).otherwise(F.col("authorZipCode"))) \
        .withColumn("authorDept", F.expr("int(authorZipCode / 1000)"))

    df.show(25)
    df.printSchema()

    logging.warning(f"{df.count()} rows updated.")

    df.write.format("delta") \
        .mode("overwrite") \
        .save("/tmp/delta_grand_debat_events")

if __name__ == "__main__":
    # Creates a session on a local master
    # To use Databricks Delta Lake, we should add delta core packages to SparkSession
    spark = SparkSession.builder \
        .appName("Ingestion the 'Grand Debate' files to Delta Lake") \
        .config("spark.jars.packages", "io.delta:delta-core_2.11:0.7.0") \
        .master("local[*]").getOrCreate()

    # setting log level, update this as per your requirement
    spark.sparkContext.setLogLevel("warn")

    main(spark)
    spark.stop()




